Battlefields: Move BF scripts out of game

Carbenium opened the pull request:
<!--- (**********************************)
      (** Fill in the following fields **)
      (**********************************) --->

**Changes proposed:**

-  Add usual script interface for battlefields
-  Move Wintergrasp BF script to `scripts` project
- One step closer to enable unity builds for `game` project

**Target branch(es):** 3.3.5/master

- [x] 3.3.5
- [ ] master


**Tests performed:** (Does it build, tested in-game, etc.)
scripts load fine, in-game seems to work

Carbenium committed:
Entities/WorldObject: Add IsInWorldPvpZone to keep API similar to master

Ovahlord commented on src/server/scripts/Battlefield/battlefield_script_loader.cpp:
```diff
@@ -0,0 +1,22 @@
+
+void AddSC_BF_wintergrasp();
+
+void AddBattlefieldScripts() {
```
codestyle

Carbenium replied:
Thanks :+1:

Ovahlord commented on src/server/shared/SharedDefines.h:
```diff
@@ -3816,6 +3816,21 @@ enum class MountResult : uint32
     Ok             = 10 // never sent
 };
 
+enum WorldStates : uint32 {
```
codestyle

Carbenium replied:
Thanks :+1:

jackpoz commented on src/server/game/Entities/Object/Object.cpp:
```diff
@@ -1089,6 +1089,19 @@ void WorldObject::RemoveFromWorld()
+    {
+        case 4197: // Wintergrasp
+            return true;
+            break;
```
I wonder if coverity complains about dead code here

Carbenium replied:
Wait, why dead code? What am I missing?

mweinelt replied:
This would never reach the break statement.

Carbenium replied:
:facepalm: 

@jackpoz Only one way to find out, I guess? ;)
I took it straight from [master](https://github.com/TrinityCore/TrinityCore/blob/master/src/server/game/Entities/Object/Object.cpp#L956). Did anyone do a coverity run on it since Feb. 2017? Because then it should have shown up there.

jackpoz commented on src/server/game/Entities/Player/Player.cpp:
```diff
@@ -8575,7 +8574,7 @@ void Player::SendLoot(ObjectGuid guid, LootType loot_type)
             }
             // For wintergrasp Quests
-            else if (GetZoneId() == AREA_WINTERGRASP)
+            else if (GetZoneId() == 4197)
```
maybe keep the old AREA_WINTERGRASP instead of a magic number ?

Carbenium replied:
`AREA_WINTERGRASP` is defined in BattlefieldWG.h which is now outside the game project. I didn't want to include it back here. Since there is no central location for such defines, I opted for the "lazy" version. Suggestions welcome.

Ovahlord replied:
SharedDefines.h it is then. This way you can also drop that enum from the script project files and use the shared define instead

Carbenium replied:
Done!

jackpoz commented on src/server/scripts/Northrend/zone_wintergrasp.cpp:
```diff
@@ -18,7 +18,7 @@
 #include "Battlefield.h"
 #include "BattlefieldMgr.h"
-#include "BattlefieldWG.h"
+#include "Battlefield/BattlefieldWG.h"
```
is the folder needed ?

Carbenium replied:
Yes. Only the `scripts` folder is defined as include directory.

Carbenium committed:
Scripts/Battlefield: Move npc_wg_give_promotion_credit script to BattlefieldWG

The script references BattlefieldWG, which is not exported in dynamic builds.
Move the script, so it is in the same library.

Shauren commented on sql/updates/world/3.3.5/2020_06_xx_xx_world.sql:
```diff
@@ -0,0 +1,8 @@
+    `comment`    text     null
+) ENGINE=MyISAM DEFAULT CHARSET=utf8;
+
+INSERT INTO world.battlefield_template (TypeId, ScriptName, comment) VALUES (1, 'battlefield_wg', null);
```
dont put database names in sql files

Carbenium replied:
Thanks for the reminder!

Shauren commented on sql/updates/world/3.3.5/2020_06_xx_xx_world.sql:
```diff
@@ -0,0 +1,8 @@
+DROP TABLE IF EXISTS `battlefield_template`;
+CREATE TABLE `battlefield_template` (
+    `TypeId`     tinyint  unsigned not null,
+    `ScriptName` char(64) default '' not null,
```
use varchar, not char

Carbenium replied:
Thanks!

Carbenium committed:
Core/Battlefield: Move SendInitWorldStatesTo to Battlefield class

Carbenium committed:
Core/Misc: Create new enum for AreaIds

Move Wintergrasp BF related values there

Carbenium committed:
Battlefields: Move BF scripts out of game

This commit introduces the usual script interface
for battlefields.

Carbenium committed:
Core/Misc: Move WorldState enum to SharedDefines

Move WG battlefield world states to the global enum
